etl/src/concurrency/future.rs

```diff
@@ -94,44 +93,48 @@ where
                 ReactiveFutureState::PollInnerFuture => match this.future.as_mut().poll(cx) {
                     Poll::Ready(Ok(inner_result)) => match inner_result {
                         Ok(_) => {
-                            *this.state = ReactiveFutureState::InvokeCallback(None);
+                            *this.state = ReactiveFutureState::InvokeCallback;
                         }
                         Err(err) => {
-                            let casted_err = format!("{err}");
-
                             *this.original_error = Some(err);
-                            *this.state = ReactiveFutureState::InvokeCallback(Some(casted_err));
+                            *this.state = ReactiveFutureState::InvokeCallback;
                         }
                     },
                     Poll::Ready(Err(err)) => {
-                        // We want to handle the most common panic types.
-                        let casted_err = if let Some(s) = err.downcast_ref::<&str>() {
-                            s.to_string()
-                        } else if let Some(s) = err.downcast_ref::<String>() {
-                            s.clone()
-                        } else {
-                            "Unknown panic error".to_string()
-                        };
-
-                        *this.original_panic = Some(Box::new(err));
-                        *this.state = ReactiveFutureState::InvokeCallback(Some(casted_err));
+                        *this.original_panic = Some(err);
+                        *this.state = ReactiveFutureState::InvokeCallback;
                     }
                     Poll::Pending => {
                         return Poll::Pending;
                     }
                 },
-                ReactiveFutureState::InvokeCallback(error) => {
+                ReactiveFutureState::InvokeCallback => {
                     if this.callback_fut.is_none() {
                         let id = this.id.clone();
-                        let callback_source = this.callback_source.clone();
-                        let error = error.clone();
+                        let mut callback_source = this.callback_source.clone();
+
+                        // We clone the original error.
+                        let original_error = this.original_error.clone();
```

Comment:
Kept it simple since we do not expect errors do be frequent, thus cloning is acceptable.